core/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -361,6 +361,15 @@ public TopDocs call() throws Exception {
                 }
             }
 
+            if (collector != null) {
+                final Collector child = collector;
+                collector = new CancellableCollector(searchContext, collector);
```

Comment:
@jpountz  what do you think about this 662e1f5e3a2e9db89511749fe104e922edbb86ad?